etherman/batch_requests.go

```diff
@@ -0,0 +1,215 @@
+// this is used in simulated environments where batch requests are not supported
+func RetrieveBlockHeadersLegacy(ctx context.Context,
+	log aggkitcommon.Logger,
+	ethClient aggkittypes.BaseEthereumClienter,
+	blockNumbers map[uint64]struct{},
+	maxConcurrency int) (map[uint64]*aggkittypes.BlockHeader, error) {
+	return retrieveBlockHeadersInBatchParallel(
+		log,
+		func(blocks map[uint64]struct{}) (map[uint64]*aggkittypes.BlockHeader, error) {
+			result := make(map[uint64]*aggkittypes.BlockHeader, len(blocks))
+			for blockNumber := range blocks {
+				header, err := ethClient.HeaderByNumber(ctx, big.NewInt(int64(blockNumber)))
+				if err != nil {
+					return nil, fmt.Errorf("RetrieveBlockHeadersLegacy: cannot get block header for block %d: %w",
+						blockNumber, err)
+				}
+				result[blockNumber] = aggkittypes.NewBlockHeaderFromEthHeader(header)
+			}
+			return result, nil
+		}, blockNumbers, 1, maxConcurrency)
+}
+
+// retrieveBlockHeadersInBatch retrieves block headers for the given block numbers using batch requests
+func retrieveBlockHeadersInBatch(ctx context.Context,
+	log aggkitcommon.Logger,
+	rpcClient aggkittypes.RPCClienter,
+	blockNumbers map[uint64]struct{},
+) (map[uint64]*aggkittypes.BlockHeader, error) {
+	if len(blockNumbers) == 0 {
+		return make(map[uint64]*aggkittypes.BlockHeader), nil
+	}
+	headers := make(map[uint64]*blockRawEth)
+	timeTracker := aggkitcommon.NewTimeTracker()
+	timeTracker.Start()
+	batch := make([]rpc.BatchElem, 0, len(blockNumbers))
+	for blockNumber := range blockNumbers {
+		headers[blockNumber] = &blockRawEth{}
+		bn := fmt.Sprintf("0x%x", blockNumber)
+		batch = append(batch, rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{bn, false},
+			Result: headers[blockNumber],
+		})
+	}
+
+	err := rpcClient.BatchCallContext(ctx, batch)
+	timeTracker.Stop()
+	if err != nil {
+		return nil, fmt.Errorf("retrieveRPCBlockHeadersInBatch(%d): BatchCallContext error: %w", len(blockNumbers), err)
+	}
+	for i, elem := range batch {
+		if elem.Error != nil {
+			return nil, fmt.Errorf("retrieveRPCBlockHeadersInBatch(%d): batch element %d (%v) error: %w", len(blockNumbers), i,
+				elem.Args,
+				elem.Error)
+		}
+	}
+	log.Debugf("retrieveRPCBlockHeadersInBatch: Retrieved block headers for blocks %d in %s (elapsed)",
+		len(blockNumbers), timeTracker.Duration().String())
+	return convertMapBlockRawEth(headers)
+}
+
+// retrieveBlockHeadersInBatchParallel split request into chuncks and execute it in parallel
+func retrieveBlockHeadersInBatchParallel(
+	logger aggkitcommon.Logger,
+	funcRetrieval func(map[uint64]struct{}) (map[uint64]*aggkittypes.BlockHeader, error),
+	blockNumbers map[uint64]struct{},
+	chunckSize, maxConcurrency int) (map[uint64]*aggkittypes.BlockHeader, error) {
+	results := make(map[uint64]*aggkittypes.BlockHeader)
+	errs := make([]error, 0)
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	sem := make(chan struct{}, maxConcurrency)
```

Comment:
You don't need this semaphore pattern to limit active go routines. Newer `Go` versions support an `errorGroup` with limit set of active go routines. You just need to pass a context to the `retrieveBlockHeadersInBatchParallel`. Something like this:

```go
	g, _ := errgroup.WithContext(ctx)
	g.SetLimit(maxConcurrency) // Built-in concurrency limiting!

	for _, chunk := range chunks {
		g.Go(func() error {
			headers, err := funcRetrieval(chunk)
			if err != nil {
				return fmt.Errorf("RetrieveBlockHeadersInBatchParallel: %w", err)
			}

			mu.Lock()
			for blockNumber, header := range headers {
				results[blockNumber] = header
			}
			mu.Unlock()
			return nil
		})
	}

	if err := g.Wait(); err != nil {
		return results, err
	}
```